src/Shared/FileMatcher.cs

```diff
@@ -1391,8 +1396,45 @@ internal static string[] GetFiles
             IEnumerable<string> excludeSpecsUnescaped = null
         )
         {
-            string[] files = GetFiles(projectDirectoryUnescaped, filespecUnescaped, excludeSpecsUnescaped, s_defaultGetFileSystemEntries, s_defaultDirectoryExists);
-            return files;
+            if (Traits.Instance.MSBuildCacheFileEnumerations)
+            {
+                string filesKey = $"{projectDirectoryUnescaped}{filespecUnescaped}";
+                string[] files;
+                if (!s_cachedFileEnumerations.Value.TryGetValue(filesKey, out files))
+                {
+                    object locks = s_cachedFileEnumerationsLock.Value.GetOrAdd(filesKey, _ => new object());
+                    lock (locks)
+                    {
+                        if (!s_cachedFileEnumerations.Value.TryGetValue(filesKey, out files))
+                        {
+                            files =
+                                s_cachedFileEnumerations.Value.GetOrAdd(
+                                    filesKey,
+                                    (_) =>
+                                        GetFiles(
+                                            projectDirectoryUnescaped,
+                                            filespecUnescaped,
+                                            excludeSpecsUnescaped,
+                                            s_defaultGetFileSystemEntries,
+                                            s_defaultDirectoryExists));
+                        }
+                    }
+                }
+
+                var filesToReturn = new string[files.Length];
+                Array.Copy(files, filesToReturn, files.Length);
```

Comment:
We discussed this a bit. Going to sort the array and return an immutable version (of course not `ImmutableArray<string>` because `MSBuildTaskHost`, but something like that).